netbox/generate_functions.go

```diff
@@ -0,0 +1,203 @@
+		ServiceName:    *serviceName,
+		Endpoint:       *endpoint,
+		Service:        *service,
+		ListOpts:       !*withoutListOpts,
+	})
+
+	// go fmt
+	res, err := format.Source(b.Bytes())
+	if err != nil {
+		b.WriteTo(os.Stderr)
+		log.Fatal(err)
+	}
+
+	f, err := os.Create(fmt.Sprintf("%s_%s.go", *endpoint, *service))
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+	f.Write(res)
+}
+
+var functionsTemplate = template.Must(template.New("").Parse(`// Copyright 2017 The go-netbox Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// GENERATED BY THE COMMAND ABOVE; DO NOT EDIT
+// This file was generated by robots at
+// {{ .Timestamp }}
+
+package netbox
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+)
+
+// {{ .ServiceName }} is used in a Client to access NetBox's {{ .Endpoint }}/{{ .Service }} API methods.
+type {{ .ServiceName }} struct {
+	c *Client
+}
+
+// Get retrieves an {{ .TypeName }} object from NetBox by its ID.
+func (s *{{ .ServiceName }}) Get(id int) (*{{ .TypeName }}, error) {
+	req, err := s.c.NewRequest(
+		http.MethodGet,
+		fmt.Sprintf("api/{{ .Endpoint }}/{{ .Service }}/%d/", id),
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	t := new({{ .TypeName }})
+	err = s.c.Do(req, t)
+	if err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+// List returns a Page associated with an NetBox API Endpoint.
+{{ if .ListOpts -}}
+func (s *{{ .ServiceName }}) List(options *List{{ .TypeName }}Options) *Page {
+	return NewPage(s.c, "api/{{ .Endpoint }}/{{ .Service }}/", options)
+}
+{{ else -}}
+func (s *{{ .ServiceName }}) List() *Page {
+	return NewPage(s.c, "api/{{ .Endpoint }}/{{ .Service }}/", nil)
+}
+{{ end }}
+
+// Extract retrives a list of {{ .TypeName }} objects from page.
+func (s *{{ .ServiceName }}) Extract(page *Page) ([]*{{ .TypeName }}, error) {
+	if err := page.Err(); err != nil {
+		return nil, err
+	}
+
+	var groups []*{{ .TypeName }}
+	if err := json.Unmarshal(page.data.Results, &groups); err != nil {
+		return nil, err
+	}
+	return groups, nil
+}
+
+// Create creates a new {{ .TypeName }} object in NetBox and returns the new object.
+func (s *{{ .ServiceName }}) Create(data *{{ .TypeName }}) (*{{ .TypeName }}, error) {
+	req, err := s.c.NewJSONRequest(http.MethodPost, "api/{{ .Endpoint }}/{{ .Service }}/", nil, data)
+	if err != nil {
+		return nil, err
+	}
+
+	g := new({{ .UpdateTypeName }})
+	err = s.c.Do(req, g)
+	if err != nil {
+		return nil, err
+	}
+	data.ID = g.ID
```

Comment:
This assumes that data is otherwise unchanged on the server, which may not be the case. For example, if you submit a device with no status set, the returned device will have `"status": 1`. I'm not sure what the main use case for the returned structure is? Depending on that, including all the default values that are returned could be important. Another option is to return only the ID and expect users to call `Get(id)` to see the full new state.

There are also types conversions that could happen, with numeric identifiers in strings (e.g. `"device_type": "1"` becoming `"device_type": 1` in the response) but that shouldn't be possible here with all the static types for input structs.